com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/MaterialBuilder.java

```diff
@@ -0,0 +1,223 @@
+import com.dynamo.bob.util.MurmurHash;
+
+@BuilderParams(name = "Material", inExts = {".material"}, outExt = ".materialc")
+public class MaterialBuilder extends Builder<Void>  {
+
+    private static final String TextureArrayFilenameVariantFormat = "_max_pages_%d.%s";
+
+    private class ShaderProgramBuildContext {
+        public String                       buildPath;
+        public String                       projectPath;
+        public IResource                    resource;
+        public ES2ToES3Converter.ShaderType type;
+        public ShaderDesc                   desc;
+
+        // Variant specific state
+        public boolean hasVertexArrayVariant;
+        public String[] arraySamplers = new String[0];
+    }
+
+    private ShaderDesc getShaderDesc(IResource resource, ShaderProgramBuilder builder, ES2ToES3Converter.ShaderType shaderType) throws IOException, CompileExceptionError {
+        builder.setProject(this.project);
+        Task<ShaderPreprocessor> task = builder.create(resource);
+        return builder.getCompiledShaderDesc(task, shaderType);
+    }
+
+    private ShaderDesc.Language findTextureArrayShaderLanguage(ShaderDesc shaderDesc) {
+        ShaderDesc.Language selected = null;
+        for (int i=0; i < shaderDesc.getShadersCount(); i++) {
+            ShaderDesc.Shader shader = shaderDesc.getShaders(i);
+            if (VariantTextureArrayFallback.isRequired(shader.getLanguage())) {
+                assert(selected == null);
+                selected = shader.getLanguage();
+            }
+        }
+
+        return selected;
+    }
+
+    private void applyVariantTextureArray(MaterialDesc.Builder materialBuilder, ShaderProgramBuildContext ctx, String inExt, String outExt) throws IOException, CompileExceptionError {
+
+        ShaderDesc.Language shaderLanguage = findTextureArrayShaderLanguage(ctx.desc);
+        if (shaderLanguage == null) {
+            return;
+        }
+
+        String shaderInputSource = new String(ctx.resource.getContent());
+
+        int maxPageCount = materialBuilder.getMaxPageCount();
+
+        // Taken from ShaderProgramBuilder.java
+        boolean isDebug = (this.project.hasOption("debug") || (this.project.option("variant", Bob.VARIANT_RELEASE) != Bob.VARIANT_RELEASE));
+        Common.GLSLCompileResult variantCompileResult = ShaderProgramBuilder.buildGLSLVariantTextureArray(shaderInputSource, ctx.type, shaderLanguage, isDebug, maxPageCount);
+
+        // No array samplers, we can use original source
+        if (variantCompileResult.arraySamplers.length == 0) {
+            return;
+        }
+
+        ShaderProgramBuilder.ShaderBuildResult variantBuildResult = ShaderProgramBuilder.makeShaderBuilderFromGLSLSource(variantCompileResult.source, shaderLanguage);
+
+        // JG: AAaah this should not be here, but we need to know of the parsed array samplers for building the indirection map..
+        variantBuildResult.shaderBuilder.setVariantTextureArray(true);
```

Comment:
Actually, this *is* needed for the engine to be able to select the proper shader when creating the program so the comment should be removed